### PR TITLE
#184637252 Reset user password via email

### PR DESCRIPTION
**Pull request by rwamugema**
### What does this PR do?

Given a user forgets the password for his/her account:

1. Should be able to make a request to reset the password.
2. Should receive an email containing a link to reset the password.

### Description of Task to be completed?

 * Design an endpoint and a controller to receive the request from the user and send an email with `link` to reset the password. 
 * Design an endpoint and a controller to verify and reset user password.

### How should this be manually tested?

**Installation**

**1. Clone the repository by running**

`git clone https://github.com/atlp-rwanda/falcons-ec-be.git
`
**2. Navigate to the project directory by running cd falcons-ec-be.**

`cd falcons-ec-be`

**5. Install the necessary dependencies by running npm install.**

`npm install`

6. **Run the development server by running npm run dev. This will start the server using Nodemon, which will automatically reload the server when changes are made to the code.**

`npm run dev`

7. **Run the tests by running npm run test**

`npm run test`

### Any background context you want to provide?

when running this on local server put into consideration the below information 👇

  * Add the the user log in token in headers with a key of `Authorization `and make sure that the value is started by `Bearer
`
* Add the token generated when reset password made into parameters of the endpoint  `PATCH:/api/v1/users/:token`

 * For making a request to reset password ,use endpoint POST: /api/v1/users/password
 * for making a reset password process, use endpoint PATCH:/api/v1/users/:token
 
kindly consider adding request body in this format while making a request to reset the password:

```
{
"email":"youremail@example.com",
}
```

kindly consider adding request body in this format while reseting the  password:
```
{
"password":"12345"
}

```
### What are the relevant pivotal tracker/Trello stories?

Tracker: 184637252

### Screenshots (if appropriate)

none

### Questions:

none

### CODE REVIEW

**Review by dean-daryl — APPROVED** [review]

File: src/controllers/userController.js [hunk 1]
```diff
@@ -100,28 +106,87 @@ const createNewUser = async (req, res) => {
...
+      subject: 'Password Reset',
+      html: `<p><p><strong style="font-size: 18px">Hello</strong> ${req.body.email}, </p><br>
+      <b>A request has been recieved to change password for your E-commerce account</b>
+      <br></p><p>Click the link below to reset your email <br>
+      ${link}
+      <br>
+      <br>
+      <span>If you did not initiate this request, ignore this message</span>
+      </p>`
+    });
```
Comment (fab-ryan) [inline src/controllers/userController.js]: use helper function to send and getting message

**Review by fab-ryan — CHANGES_REQUESTED** [review]
do requested changes

**Review by kirengaboris — APPROVED** [review]

File: src/helpers/sendMessage.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import sgMail from '@sendgrid/mail';
+import dotenv from 'dotenv';
+
+dotenv.config();
+sgMail.setApiKey(process.env.API_KEY);
+const { SENDGRID_EMAIL } = process.env;
+const sendMessage = async (userEmail, token, clientURL) => {
+  try {
```
Comment (fab-ryan) [inline src/helpers/sendMessage.js]: use meaningful variable because we will have a lot of sending messages variables

**Review by fab-ryan — CHANGES_REQUESTED** [review]
do that request and I will merged it

**Review by fab-ryan — APPROVED** [review]

**Pull request merged**